lambdas/dla-migration/src/index.ts

```diff
@@ -1,9 +1,162 @@
+ *
+ * @param targetPath
+ * @returns whether the innermost folder of the filePath appears to be a timestamp
+ */
+export const identifyDatedPath = (targetPath: string): boolean => (
+  moment(path.basename(path.dirname(targetPath)), 'YYYY-MM-DD').isValid()
+);
+
+/**
+ * Manipulate target filepath to put file is YYYY-MM-DD sub-folder
+ *
+ * @param targetPath
+ * @param message - DLARecord message from which timestamp can be extracted
+ * @returns targetPath with lowest 'folder' added in form YYYY-MM-DD
+ */
+export const addDateIdentifierToPath = (targetPath: string, message: DLARecord): string => (
+  path.join(
+    path.dirname(targetPath),
+    extractDateString(message),
+    path.basename(targetPath)
+  )
+);
+
+/**
+ * pull S3 Object from sourcePath, update it to new DLA structure and push it to targetPath
+ * noop if skip is true and an object already exists at targetPath
+ *
+ * @param bucket
+ * @param sourcePath
+ * @param targetPath
+ * @param skip - skip if targetPath already exists
+ * @returns whether the logic was actually run (not skipped)
+ */
+export const updateDLAFile = async (
+  bucket: string,
+  sourcePath: string,
+  targetPath: string
+): Promise<boolean> => {
+  const inDateForm = identifyDatedPath(targetPath);
+
+  logger.info(`About to process ${sourcePath}`);
+  const dlaObject = await getJsonS3Object(bucket, sourcePath);
+  const hoisted = await hoistCumulusMessageDetails(dlaObject);
+  let massagedTargetPath: string;
+  if (!inDateForm) {
+    massagedTargetPath = addDateIdentifierToPath(targetPath, hoisted);
+  } else {
+    massagedTargetPath = targetPath;
+  }
+
+  await putJsonS3Object(bucket, massagedTargetPath, hoisted);
+  logger.info(`Migrated file from bucket ${bucket}/${sourcePath} to ${massagedTargetPath}`);
+  if (massagedTargetPath !== sourcePath) {
+    await deleteS3Object(bucket, sourcePath);
+    logger.info(`Deleted file ${bucket}/${sourcePath}`);
+  }
+  return true;
+};
+
+/**
+ * update a batch of DLA files under prefix and push them to the targetDirectory
+ * skip files that appear to already have been processed if skip is set to true
+ *
+ * @param bucket
+ * @param targetDirectory
+ * @param prefix
+ * @param skip - skip files that are already present at the target directory, default false
+ */
+export const updateDLABatch = async (
+  bucket: string,
+  targetDirectory: string,
+  sourceDirectory: string
+): Promise<Array<boolean>> => {
+  const out = [];
+  const sourceDir = manipulateTrailingSlash(sourceDirectory, true);
+  const targetDir = manipulateTrailingSlash(targetDirectory, true);
+  for await (
+    const objectBatch of listS3ObjectsV2Batch({ Bucket: bucket, Prefix: sourceDir })
+  ) {
+    const validKeys = objectBatch.map((obj) => obj.Key);
+    const targetPaths = validKeys.map(
+      (filePath) => filePath.replace(
+        sourceDir,
+        targetDir
+      )
+    );
+
+    const zipped: Array<[string, string]> = zip(validKeys, targetPaths) as Array<[string, string]>;
```

Comment:
I'm confused with the logic here.  Can we add logic for processing each key?

Reply:
the function takes a sourceFile and a targetFile, and this joins them to each other so we can map over them